### ISSUE TITLE
Research transparent HTTP proxy for authN and authZ

### ISSUE THREAD

**Issue by arunabha** [mesosup/roger-mesos#1]
Ideally we need an HTTP proxy to put in front of existing Mesos frameworks, to enable generic authentication and authorization. 

An ideal solution will  

1. Be completely transparent baring any authentication related flows.
2. Provide pluggable authentication and authorization backends.
3. Provide out of the box integration with major SSO providers.
4. Be programmable, so default behaviour can be changed if necessary.

**Comment by rvenkatesh25** [comment]
There are several options for transparent proxies that support programmable plugins - nginx, apache traffic server, varnish, haproxy (which is already used in this ecosystem). However, there are very few options for proxies that support out-of-the-box SSO support. 

The most popular seems to be [oauth2_proxy](https://github.com/bitly/oauth2_proxy) - and it is pretty actively worked upon (with recent commits dating 12/15/2015)

It's written in Go and has pre-defined SSO integration with google, github and linkedin. And it is extensible as well.

Here's a quick tutorial on how to get it to work:

1. Setup a vagrant node and start mesos and marathon on it. Follow tutorials 1 through 4 [here](https://open.mesosphere.com/advanced-course/starting-marathon/)

2. After step 1, I'm assuming marathon is up and running on http://192.168.33.10:8080/ 

3. Install oauth2_proxy
```bash
$ export GOPATH=$HOME/go
$ export PATH=$PATH:$GOPATH/bin
$ go get github.com/bitly/oauth2_proxy
$ which oauth2_proxy
~/go/bin/oauth2_proxy
```

4. For this tutorial, I'm integrating with github SSO

5. Go to https://github.com/settings/applications/new and register a new application with details - Name: authn-test, Callback URL: http://192.168.33.10:4180/oauth2/callback

6. Created https://github.com/settings/applications/278542 (if it is not accessible create a new one)

7. Create the following config file
```bash
$ cat oauth2_proxy.cfg
## <addr>:<port> to listen on for HTTP/HTTPS clients
http_address = "192.168.33.10:4180"

## the OAuth Redirect URL.
redirect_url = "http://192.168.33.10:4180/oauth2/callback"

## the http url(s) of the upstream endpoint. If multiple, routing is based on path
upstreams = [
    "http://192.168.33.10:8080/"
]

## pass HTTP Basic Auth, X-Forwarded-User and X-Forwarded-Email information to upstream
pass_basic_auth = true

## pass the request Host Header to upstream
## when disabled the upstream Host is used as the Host Header
# pass_host_header = true

## Email Domains to allow authentication for (this authorizes any email on this domain)
## for more granular authorization use `authenticated_emails_file`
## To authorize any email addresses use "*"
email_domains = [
     "*"
]

## The OAuth Client ID, Secret
client_id = "f3e816592e3b3bcc503f"
client_secret = "3b18a02961cf89f4cda29ed8649404addc320c3e"

## Cookie Settings
## Name     - the cookie name
## Secret   - the seed string for secure cookies; should be 16, 24, or 32 bytes
##            for use with an AES cipher when cookie_refresh or pass_access_token
##            is set
## Domain   - (optional) cookie domain to force cookies to (ie: .yourcompany.com)
## Expire   - (duration) expire timeframe for cookie
## Refresh  - (duration) refresh the cookie when duration has elapsed after cookie was initially set.
##            Should be less than cookie_expire; set to 0 to disable.
##            On refresh, OAuth token is re-validated.
##            (ie: 1h means tokens are refreshed on request 1hr+ after it was set)
## Secure   - secure cookies are only sent by the browser of a HTTPS connection (recommended)
## HttpOnly - httponly cookies are not readable by javascript (recommended)
cookie_name = "_oauth2_proxy"
cookie_secret = "random_secret_16"
cookie_domain = ""
cookie_expire = "168h"
cookie_refresh = "8h"
cookie_secure = false
cookie_httponly = true

provider = "github"
```

8. Start the proxy
```bash
oauth2_proxy -config=./oauth2_proxy.cfg
```

9. Now access http://192.168.33.10:4180/ - it redirects to github for auth. Each link on the Marathon and Chronos UI is relative, so every click goes through the proxy. 

10. The proxy sends the X-Forwarded-User and some other integration specific headers to the upstream app, where we can do further actions like authz and quota.

**Comment by rvenkatesh25** [comment]
1. Once we have a authorization backend, we can use a proxy chain like:
```
User -> oauth2_proxy w/ google -> oauth2_proxy/some_other_proxy w/ authz -> Marathon 
```
such that the upstream application will get the user, role and other required information in HTTP headers. 

2. I am not sure if the Marathon/Chronos UI are themselves behind a HAProxy (or is HAProxy only used to load balance the apps that run on these frameworks). 

If it is the former, we can do it in either permutation:
```
haproxy -> oauth2_proxy -> Marathon
or
oauth2_proxy -> haproxy -> Marathon
```

**Comment by arunabha** [comment]
Excellent work. Thanks, Venky !

**Comment by rvenkatesh25** [comment]
Other alternatives:

1. Nginx with [oauth2 Lua module](http://chairnerd.seatgeek.com/oauth-support-for-nginx-with-lua/) - Not sure if this is actively maintained though. [OpenResty](http://openresty.org/) has regular releases, and it includes the oauth2 lua module mentioned before, but it's an overkill (it's actually meant to be a nginx + a big collection of 3rd party modules)

Anyway, for nginx  it's just better to use oauth2_proxy with its [auth_request](http://nginx.org/en/docs/http/ngx_http_auth_request_module.html) hook

2. Apache httpd with [mod_auth_openidc](https://github.com/pingidentity/mod_auth_openidc) and [mod_proxy](http://httpd.apache.org/docs/2.2/mod/mod_proxy.html) - mod_proxy is a part of standard distribution, mod_auth_openidc is actively maintained. But in general, using apache httpd just to use mod_proxy is heavy weight

ATS and Varnish are more caching oriented, didn't see any oauth plugins for them. Likewise for HAProxy. 

So for Oauth, using oauth2_proxy created by [bitly](https://bitly.com/) with proxy chaining seems to be a good option. I am also considering OAuth to be the primary SSO type to support for now.  

Regarding support for other forms of SSO, AD/Kerberos based SSO support is very thin for any of these proxies. We might have to write something on our own. Many enterprises may have this type of SSO.

SAML v2 based SSO is another common method, with several public providers (google, linkedin, yahoo). I will take a look at proxy support for saml as well.

**Comment by arunabha** [comment]
One issue I can see with oauth2_proxy (and I guess for any OAuth SSO) is
that we'll need to register the OAuth callback with the OAuth provider.
This obviously needs to be done on a per install basis, so we'll need to
provide the instructions in the docs and hope that people don't screw up :-(

On a side note, how did you get the 192.x.x.x callback address to work with
Github ?

**Comment by rvenkatesh25** [comment]
That's true, any SSO will involve installation specific integration instructions. What we can do is package/publish this with an unrestrictive "demo" app integration, say anyone with a google account is allowed. And also provide instructions to customize the SSO integration. 

re: "On a side note, how did you get the 192.x.x.x callback address to work with Github?"
The callback url is returned in a Location header with a 302 response from the Oauth service provider, which is followed by the browser. So 192.x.x.x or even localhost works.

**Comment by arunabha** [comment]
Hmm, the 302 is pretty ingenious, does Google do that ?

The Nginx + oauth2_proxy sounds like the right combination for now ? That
way we can select between HTTP basic auth or SSO using OAuth based on Nginx
configuration.

**Comment by ankanm** [comment]
This is very promising. 
@rvenkatesh25 - Have you looked at vault and it's [auth backends](https://vaultproject.io/docs/auth/)?

**Comment by rvenkatesh25** [comment]
@ankanm I hadn't looked at the Vault before, checked it out only after you mentioned it. 

IMO it's not directly applicable to our use case - vault is a software to store and retrieve secrets (keys, passwords), whereas we are looking for an overlay that can do the auth functions. 

One take away is their design to have pluggable auth method to access the secrets stored in vault, and have a single interface to be plugged with github, ldap etc. It would be a great if some transparent proxy did the same (which so far I haven't been able to find). Certainly defines a good sub-project scope - we can write one such proxy. 

What did you have in mind? Can we use Vault in some other way here?
